tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,372 @@
+
+            """
+            return layer["conv1"](in_tensor)
+
+        def spatially_constrained_layer1(
+            layer: torch.nn.Module,
+            in_tensor: torch.Tensor,
+            out_height: int = 13,
+            out_width: int = 13,
+        ) -> tuple:
+            """Spatially constrained layer 1.
+
+            Estimates row, column and height for
+            spatially_constrained_layer2 layer mapping.
+
+            Args:
+                layer (torch.nn.Module):
+                    Torch layer as ModuleDict.
+                in_tensor (torch.Tensor):
+                    Input Tensor.
+                out_height (int):
+                    Output height.
+                out_width (int):
+                    Output Width
+
+            Returns:
+                tuple of :class:`torch.Tensor`:
+                    Row, Column and height estimates used for
+                    spatially_constrained_layer2 mapping.
+
+            """
+            sigmoid = layer["conv1"](in_tensor)
+            sigmoid0 = sigmoid[:, 0:1, :, :] * (out_height - 1)
+            sigmoid1 = sigmoid[:, 1:2, :, :] * (out_width - 1)
+            sigmoid2 = sigmoid[:, 2:3, :, :]
+            return sigmoid0, sigmoid1, sigmoid2
+
+        def spatially_constrained_layer2(network, sc1_0, sc1_1, sc1_2) -> torch.Tensor:
+            """Spatially constrained layer 2.
+
+            Estimates row, column and height for sc2 layer mapping.
+
+            Args:
+                network (:class:`.SCCNN`):
+                    An initiated SCCNN class.
+                sc1_0 (torch.Tensor):
+                    Output of spatially_constrained_layer1 estimating
+                    the x position of the nucleus.
+                sc1_1 (int):
+                    Output of spatially_constrained_layer1 estimating
+                    the y position of the nucleus.
+                sc1_2 (int):
+                    Output of spatially_constrained_layer1 estimating
+                    the confidence in nucleus detection.
+
+            Returns:
+                :class:`torch.Tensor`:
+                    Probability map using the estimates from
+                    spatially_constrained_layer1.
+
+            """
+            x = torch.tile(
+                network.x, dims=[sc1_0.size(0), 1, 1, 1]
+            )  # Tile for batch size
+            y = torch.tile(network.y, dims=[sc1_0.size(0), 1, 1, 1])
+            xvr = (x - sc1_0) ** 2
+            yvc = (y - sc1_1) ** 2
+            out_map = xvr + yvc
+            out_map_threshold = torch.lt(out_map, network.radius).type(torch.float32)
+            denominator = 1 + (out_map / 2)
+            sc2 = sc1_2 / denominator
+            return sc2 * out_map_threshold
+
+        l1 = conv_act_branch(self.layer["l1"], in_tensor=input_tensor)
+        p1 = self.layer["pool1"](l1)
+        l2 = conv_act_branch(self.layer["l2"], in_tensor=p1)
```

Comment:
If you deleted the `conv_act_branch` in `forward` function, this line and lines L276, L277, L279 can be easily re-written.

Reply:
I get this error if I use the above suggestion

self = ModuleDict(
  (conv1): Sequential(
    (0): Conv2d(3, 30, kernel_size=(2, 2), stride=(1, 1))
    (1): ReLU()
  )
)
input = (tensor([[[[0.8431, 0.8549, 0.8549,  ..., 0.8510, 0.8941, 0.9412],
          [0.8118, 0.8588, 0.8863,  ..., 0.8588, 0...., 0.7333, 0.7412,  ..., 0.5765, 0.6078, 0.6471],
          [0.7294, 0.7412, 0.7529,  ..., 0.6314, 0.6471, 0.6706]]]]),)

    def _forward_unimplemented(self, *input: Any) -> None:
        r"""Defines the computation performed at every call.
    
        Should be overridden by all subclasses.
    
        .. note::
            Although the recipe for forward pass needs to be defined within
            this function, one should call the :class:`Module` instance afterwards
            instead of this since the former takes care of running the
            registered hooks while the latter silently ignores them.
        """
>       raise NotImplementedError
E       NotImplementedError

C:\tools\Anaconda3\envs\tiatoolbox-dev\lib\site-packages\torch\nn\modules\module.py:201: NotImplementedError